src/turbomind/api/python/linear.cc

```diff
@@ -274,8 +287,38 @@ struct Linear::Impl {
     }
 
 private:
+    static gemm::Workspace& getWorkspace(int device_id, cudaStream_t stream) {
+        std::lock_guard<std::mutex> lock(cache_mutex_);
+        auto key = std::make_tuple(device_id, stream);
+        auto it = workspace_cache_.find(key);
+        if (it != workspace_cache_.end()) {
+            return *it->second;
+        }
+
+        // create a new workspace if cache missed
+        // auto workspace = std::make_shared<gemm::Workspace>();
+        auto workspace = std::shared_ptr<gemm::Workspace>(new gemm::Workspace,
+            [](gemm::Workspace* p){
+                cudaFreeAsync(p->barriers, 0);
+                cudaFreeAsync(p->partials, 0);
+            });
+        workspace->barriers_size = gemm::Gemm::kBarriersSize;
+        workspace->partials_size = gemm::Gemm::kPartialsSize;
+        cudaMallocAsync(&workspace->barriers, workspace->barriers_size, stream);
```

Comment:
`check_cuda_error` on the allocations to avoid slient OOM.

Reply:
Sure, added